src/main/java/com/ryandw11/structure/bottomfill/DefaultBottomFill.java

```diff
@@ -1,77 +1,121 @@
 package com.ryandw11.structure.bottomfill;
 
+import com.google.common.base.Preconditions;
 import com.ryandw11.structure.CustomStructures;
 import com.ryandw11.structure.structure.Structure;
+import com.sk89q.worldedit.extent.clipboard.Clipboard;
+import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormat;
+import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormats;
+import com.sk89q.worldedit.extent.clipboard.io.ClipboardReader;
+import com.sk89q.worldedit.math.BlockVector2;
+import com.sk89q.worldedit.math.BlockVector3;
 import org.bukkit.Location;
 import org.bukkit.Material;
+import org.bukkit.World;
 import org.bukkit.scheduler.BukkitRunnable;
-import org.bukkit.scheduler.BukkitTask;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.Queue;
 
 /**
  * The default implementation for the bottom fill feature.
  */
-public class DefaultBottomFill extends BukkitRunnable implements BottomFillImpl {
+public class DefaultBottomFill extends BukkitRunnable implements BottomFill {
 
-    private Location minLoc;
-    private Location maxLoc;
-    private Location location;
     private Structure structure;
-    private int currentX;
-    private int currentY;
-    private int currentZ;
-    private BukkitTask bukkitTask;
+    private Location spawnLocation;
+    private Queue<BlockVector2> groundPlane;
+    private int minY;
+    private Material fillMaterial;
 
     @Override
-    public void performFill(Structure structure, Location spawnLocation, Location minLoc, Location maxLoc) {
-        if (structure.getBottomSpaceFill().getFillMaterial(spawnLocation.getBlock().getBiome()).isEmpty())
-            return;
+    public void performFill(Structure structure, Location spawnLocation, Location minLoc, Location maxLoc) throws IOException {
+        var fillMaterial = structure.getBottomSpaceFill().getFillMaterial(spawnLocation.getBlock().getBiome());
+        if (fillMaterial.isPresent()) {
+            this.fillMaterial = fillMaterial.get();
+        } else return;
 
         this.structure = structure;
-        this.location = spawnLocation;
-        this.minLoc = new Location(minLoc.getWorld(), minLoc.getBlockX(), minLoc.getBlockY() - 1, minLoc.getBlockZ());
-        this.maxLoc = new Location(minLoc.getWorld(), maxLoc.getBlockX(), minLoc.getBlockY() - 1, maxLoc.getBlockZ());
-        currentX = minLoc.getBlockX();
-        currentY = minLoc.getBlockY() - 1;
-        currentZ = minLoc.getBlockZ();
-        bukkitTask = this.runTaskTimer(CustomStructures.getInstance(), 0, 1);
+        this.spawnLocation = spawnLocation;
+        this.minY = minLoc.getBlockY();
+
+        // The 2D plane which the blocks will be ground placed on
+        groundPlane = new LinkedList<>();
+
+        // The world to manipulate
+        World world = spawnLocation.getWorld();
+        Preconditions.checkNotNull(world, "world");
```

Comment:
Don't use an additional library not used anywhere else in the plugin for this. Also the error message for being null should be easily understood by the average plugin users.

The check for null should probably be manually with a descriptive log message.